File: lecture_examples/javascript/3_whitespace_and_boundaries.js
// LECTURE EXAMPLE 11: leading spaces
/**
 * @function findLeadingSpaces
 * Return the input string if it starts with one or more space characters.
 * 
 * @param {string} string Input string for regex.
 * 
 * @example
 *   findLeadingSpaces('  space in the middle')
 *   // => '  space in the middle'
 * 
 *   findLeadingSpaces('no leading space')
 *   // => null
 * 
 * @returns {string} input string if regex matches; otherwise null.
 */
const findLeadingSpaces = function(string) {
  const leadingSpaceRegex = /^ +.*/;
  const match = leadingSpaceRegex.exec(string);

  if (!match) return null

  // exec returns an array whose first element is the complete match
  return match[0];
}

// LECTURE EXAMPLE 12: Match full sentences
/**
 * @function matchFullSentence
 * Return true if string represents a sentence.
 * 
 * In this case, a sentence is defined by starting with a capital
 * letter, and ending with sentence punctuation: ! . ?
 * There should be no sentence punctuation in the middle.
 *
 * @param {string} string Input string for regex.
 * 
 * @example
 *   matchFullSentence('Do you want to get coffee with me?')
 *   // => true
 * 
 *   matchFullSentence('I love kittens!')
 *   // => true
 * 
 *   matchFullSentence('This is the best Udemy course ever. Why are you laughing?')
 *   // => false
 * 
 * @returns {boolean} true if regex matches input string; otherwise false
 */
const matchFullSentence = function(string) {
  const fullSentenceRegex = /^[A-Z][^\.!?]+[\.?!]$/;
  return fullSentenceRegex.test(string);
}

// LECTURE EXAMPLE 13: full sentence with multiple punctuation
/**
 * @function matchFullSentenceMultiplePuncutation
 * Return true if string represents a sentence.
 * 
 * In this case, a sentence is defined by starting with a capital
 * letter, and ending with one or more sentence punctuation character: ! . ?
 * There should be no sentence punctuation in the middle.
 * 
 * @param {string} string Input string for regex.
 * 
 * @example
 *   matchFullSentenceMultiplePuncutation('Do you want to get coffee with me?')
 *   // => true
 * 
 *   matchFullSentenceMultiplePuncutation('I love kittens!')
 *   // => true
 * 
 *   matchFullSentenceMultiplePuncutation('Who do you think you are?!?')
 *   // => true
 * 
 *   matchFullSentenceMultiplePuncutation('This is the best Udemy course ever. Why are you laughing?')
 *   // => false
 *
 * @returns {boolean} true if regex matches input string; otherwise false
 */
const matchFullSentenceMultiplePuncutation = function(string) {
  const fullSentenceWithMultiplePuncutationRegex = /^[A-Z][^\.!?]+[\.?!]+$/;
  return fullSentenceWithMultiplePuncutationRegex.test(string);
}

module.exports = {
  findLeadingSpaces,
  matchFullSentence,
  matchFullSentenceMultiplePuncutation,
}

File: lecture_examples/python/3_whitespace_and_boundaries.py
import re


# LECTURE EXAMPLE 11: leading spaces
def find_leading_spaces(string):
    """Return the input string if it starts with one or more space characters.

    Return None otherwise.
    non-verbose regex: r'^ +.*'

    >>> find_leading_spaces('  space in the middle')
    '  space in the middle'

    >>> find_leading_spaces('no leading space')
    """
    match = re.match(
        r"""
            ^   # beginning of the string
            \ + # one or more space characters
            .*  # the rest of the string
        """,
        string,
        flags=re.VERBOSE
    )

    if match is None:
        return None

    return match[0]  # return actual match if one was found


# LECTURE EXAMPLE 12: Match full sentences
def match_full_sentence(string):
    """Return True if string represents a sentence.

    In this case, a sentence is defined by starting with a capital
    letter, and ending with sentence punctuation: ! . ?
    There should be no sentence punctuation in the middle.

    Return False otherwise.
    non-verbose regex: r'^[A-Z][^\.!?]+[\.?!]$'

    >>> match_full_sentence('Do you want to get coffee with me?')
    True

    >>> match_full_sentence('I love kittens!')
    True

    >>> match_full_sentence('This is the best Udemy course ever. Why are you laughing?')
    False
    """
    match = re.search(
        r"""
            ^         # beginning of string
            [A-Z]     # one capital letter
            [^\.!?]+ # one or more non sentence-ending punctuation characters
            [\.?!]    # one sentence-ending punctuation character
            $         # end of string
        """,
        string,
        flags=re.VERBOSE
    )

    return match is not None


# LECTURE EXAMPLE 13: full sentence with multiple punctuation
def match_full_sentence_multiple_puncutation(string):
    """Return True if string represents a sentence.

    In this case, a sentence is defined by starting with a capital
    letter, and ending with one or more sentence punctuation character: ! . ?
    There should be no sentence punctuation in the middle.

    Return False otherwise.
    non-verbose regex: r'^[A-Z][^\.!?]+[\.?!]+$'

    >>> match_full_sentence_multiple_puncutation('Do you want to get coffee with me?')
    True

    >>> match_full_sentence_multiple_puncutation('I love kittens!')
    True

    >>> match_full_sentence_multiple_puncutation('Who do you think you are?!?')
    True

    >>> match_full_sentence_multiple_puncutation('This is the best Udemy course ever. Why are you laughing?')
    False
    """
    match = re.search(
        r"""
            ^         # beginning of string
            [A-Z]     # one capital letter
            [^\.!?]+ # one or more non sentence-ending punctuation characters
            [\.?!]+   # one or more sentence-ending punctuation character
            $         # end of string
        """,
        string,
        flags=re.VERBOSE
    )

    return match is not None
